src/internal/storage/fileset/buffer.go

```diff
@@ -8,118 +8,102 @@ import (
 )
 
 type Buffer struct {
-	additive map[string]*file
-	deletive map[string]*file
+	additive map[string]map[string]*file
+	deletive map[string]map[string]*file
 }
 
 type file struct {
-	path  string
-	parts map[string]*part
-}
-
-type part struct {
-	tag string
-	buf *bytes.Buffer
+	path string
+	tag  string
+	buf  *bytes.Buffer
 }
 
 func NewBuffer() *Buffer {
 	return &Buffer{
-		additive: make(map[string]*file),
-		deletive: make(map[string]*file),
+		additive: make(map[string]map[string]*file),
+		deletive: make(map[string]map[string]*file),
 	}
 }
 
-func (b *Buffer) Add(p, tag string) io.Writer {
-	p = Clean(p, false)
-	if _, ok := b.additive[p]; !ok {
-		b.additive[p] = &file{
-			path:  p,
-			parts: make(map[string]*part),
-		}
+func (b *Buffer) Add(path string, tag ...string) io.Writer {
+	path = Clean(path, false)
+	if _, ok := b.additive[path]; !ok {
+		b.additive[path] = make(map[string]*file)
 	}
-	buf := &bytes.Buffer{}
-	b.additive[p].parts[tag] = &part{
-		tag: tag,
-		buf: buf,
+	taggedFiles := b.additive[path]
+	var fileTag string
+	if len(tag) > 0 {
+		fileTag = tag[0]
 	}
-	return buf
+	if _, ok := taggedFiles[fileTag]; !ok {
+		taggedFiles[fileTag] = &file{
+			path: path,
+			tag:  fileTag,
+			buf:  &bytes.Buffer{},
+		}
+	}
+	f := taggedFiles[fileTag]
+	return f.buf
 }
 
-func (b *Buffer) Delete(p string, tag ...string) {
-	p = Clean(p, IsDir(p))
-	if IsDir(p) {
+func (b *Buffer) Delete(path string, tag ...string) {
+	path = Clean(path, IsDir(path))
+	if IsDir(path) {
 		// TODO: Linear scan for directory delete is less than ideal.
 		// Fine for now since this should be rare and is an in-memory operation.
 		for file := range b.additive {
-			if strings.HasPrefix(file, p) {
+			if strings.HasPrefix(file, path) {
 				delete(b.additive, file)
 			}
 		}
-	}
-	if len(tag) == 0 {
-		delete(b.additive, p)
-		b.deletive[p] = &file{
-			path: p,
-		}
 		return
 	}
-	if file, ok := b.additive[p]; ok {
-		delete(file.parts, tag[0])
+	var fileTag string
+	if len(tag) > 0 {
+		fileTag = tag[0]
 	}
-	if _, ok := b.deletive[p]; !ok {
-		b.deletive[p] = &file{
-			path:  p,
-			parts: make(map[string]*part),
-		}
+	if taggedFiles, ok := b.additive[path]; ok {
+		delete(taggedFiles, fileTag)
+	}
+	if _, ok := b.deletive[path]; !ok {
+		b.deletive[path] = make(map[string]*file)
+	}
+	taggedFiles := b.deletive[path]
+	taggedFiles[fileTag] = &file{
+		path: path,
+		tag:  fileTag,
 	}
-	b.deletive[p].parts[tag[0]] = &part{tag: tag[0]}
 }
 
 func (b *Buffer) WalkAdditive(cb func(string, string, io.Reader) error) error {
```

Comment:
Nit: fill in the parameter names or add a comment so we can determine which is tag and which is path with autocomplete.